From 3d71de3f14db75d88cacb219c98620947444df09 Mon Sep 17 00:00:00 2001
From: Luciano Pacheco <lucmult@chromium.org>
Date: Tue, 26 Apr 2022 04:25:16 +0000
Subject: [PATCH] M101: Prevent multiple SelectFileDialog in
 MediaGalleriesPermissionController

Similar to crrev.com/c/3546821

(cherry picked from commit 5d50a2dc16c935df5f44223d98fab893f2372383)

Bug: 1309583
Change-Id: I8089d6a1acc44131955e68d895a314270d55c291
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3569405
Commit-Queue: Luciano Pacheco <lucmult@chromium.org>
Auto-Submit: Luciano Pacheco <lucmult@chromium.org>
Reviewed-by: Lei Zhang <thestig@chromium.org>
Commit-Queue: Lei Zhang <thestig@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#988853}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3581508
Cr-Commit-Position: refs/branch-heads/4951@{#1055}
Cr-Branched-From: 27de6227ca357da0d57ae2c7b18da170c4651438-refs/heads/main@{#982481}
---
 .../media_galleries_permission_controller.cc          | 11 +++++++++++
 .../media_galleries_permission_controller.h           |  1 +
 2 files changed, 12 insertions(+)

--- a/chrome/browser/media_galleries/media_galleries_permission_controller.cc
+++ b/chrome/browser/media_galleries/media_galleries_permission_controller.cc
@@ -202,6 +202,10 @@
 
 // This is the 'Add Folder' button.
 void MediaGalleriesPermissionController::DidClickAuxiliaryButton() {
+  // Early return if the select file dialog is already active.
+  if (select_folder_dialog_)
+    return;
+
   base::FilePath default_path =
       extensions::file_system_api::GetLastChooseEntryDirectory(
           extensions::ExtensionPrefs::Get(GetProfile()), extension_->id());
@@ -275,6 +279,10 @@
   return web_contents_;
 }
 
+void MediaGalleriesPermissionController::FileSelectionCanceled(void* params) {
+  select_folder_dialog_.reset();
+}
+
 void MediaGalleriesPermissionController::FileSelected(
     const base::FilePath& path,
     int /*index*/,
@@ -301,6 +309,7 @@
     iter->second.selected = true;
     forgotten_galleries_.erase(gallery_id);
     dialog_->UpdateGalleries();
+    select_folder_dialog_.reset();
     return;
   }
 
@@ -311,6 +320,7 @@
         iter->second.pref_info.device_id == gallery.device_id) {
       iter->second.selected = true;
       dialog_->UpdateGalleries();
+      select_folder_dialog_.reset();
       return;
     }
   }
@@ -320,6 +330,7 @@
   // The old prefId is retained for blocklisted galleries.
   gallery.pref_id = GetDialogId(gallery.pref_id);
   new_galleries_[gallery.pref_id] = Entry(gallery, true);
+  select_folder_dialog_.reset();
   dialog_->UpdateGalleries();
 }
 
--- a/chrome/browser/media_galleries/media_galleries_permission_controller.h
+++ b/chrome/browser/media_galleries/media_galleries_permission_controller.h
@@ -123,6 +123,7 @@
   void FileSelected(const base::FilePath& path,
                     int index,
                     void* params) override;
+  void FileSelectionCanceled(void* params) override;
 
   // RemovableStorageObserver implementation.
   // Used to keep dialog in sync with removable device status.
